libsrc/apple2/acc_detect_fastchip.s

```diff
@@ -0,0 +1,46 @@
+;
+; Marco van den Heuvel, 2018-05-04
+;
+
+; unsigned char detect_fastchip (void);
+;
+;/* Check for the presence of the Apple IIe Fast Chip slot card.
+; *
+; * Possible return values:
+; * 0x00  : Fast Chip slot card not present
+; * 0x01  : Fast Chip slot card present
+; */
+
+        .export         _detect_fastchip
+        .import         _get_ostype
+
+        .include        "accelerator.inc"
+
+_detect_fastchip:
+        jsr     _get_ostype
+        ldx     #$00
+        cmp     #$30                ; Apple IIe
+        beq     check_slot
+        cmp     #$31                ; Apple IIe enhanced
+        beq     check_slot
+        cmp     #$40                ; Apple IIe Option Card
+        bne     not_found
+
+check_slot:
+        ldy     #$04                ; loop 4 times
+        lda     #FASTCHIP_UNLOCK    ; load the unlock value
+unlock_loop:
+        sta     FASTCHIP_LOCK_REG   ; store in lock register
+        dey
+        bne     unlock_loop
+        sta     FASTCHIP_ENABLE_REG ; enable the Fast Chip
+        ldy     FASTCHIP_ENABLE_REG ; bit 7 will be high when enabled
```

Comment:
I'm unsure, but the detection code from the user manual is as follows:

LDY #$6A ; Load Unlock Value
STY $C06A ; Send to the FASTChip //e
STY $C06A ;
STY $C06A ;
STY $C06A ;
STY $C06B ; Enable FASTChip //e
LDY $C06B ; Is FASTChip //e enabled?
BMI EXIST ; Jmp if FASTChip //e exist.

So it's completely based on that.

Reply:
You are aware that your code changes the timing of the writes to $C06A. Are you sure that this is not harmful?
Do you have hardware to test the code? IMO it's a mood point to write accelerators for hardware you don't have. It's a gamble whether it will word on the real thing...
Just my 2¢

Reply:
I must admit that I fully agree with @groessler:

- Replacing the four consecutive absolute writes with a loop seems to indicate that you are not experienced with the way such things tend to work in the Apple II.

- Your remark "So it's completely based on that." seems to indicate that you don't have a FastChipIIe available for testing. At least as far as I know we have never merged untested hardware-related code to cc65 so far.

If there would be a clear, precise documentation and/or known-to-work example program(s) for the hardware in question one might think about merging hardware-ralted code nevertheless but I certainly don't see this with the FastChipIIe :-(

In contrast http://www.a2heaven.com/webshop/resources/pdf_document/18/82/c.pdf section 'FASTChip //e Programmer's Reference' and section 'Programming Hints' contradict each other in my opinion. Especially the term `unlocked` vs `enabled` are ambiguous. Intuitively I'd say that `unlocked` means that the configuration registers are generally protected from accidental change by requiring to be unlocked first. But then one would lock them after being done with changing the configuration. But this perspective doesn't match the 'Programming Hints' :-(

Reply:
Ok, if untested hardware code is not accepted then I'm done. I don't even have a c64 or c128, so there will be no further contributions from me.

Reply:
Okay, sorry if again have no clue but my understanding is that your are a long-term contributor to VICE. And my understanding is that VICE emulates all the hardware you have contributed code for to cc65. And that you have tested your contributions with those VICE emulations. Isn‘t that correct?

My position: If there‘s no code at all for a given hardware in cc65 then code only tested with „some“ popular emulation maybe has issues with the real hardware but after all it works at least for others using the same emulation - and it‘s for sure better than nothing. If however, someone want‘s to change/extend/fix existing hardware-related code in cc65 it should be tested on real hardware to avoid regressions on real hardware.

To my knowledge there‘s no FastChipIIe emulation available and (as said) the available docs are rather bad and there‘s no example code and there‘s no hardware schematics / firmware soure code available. So your FastChipIIe contribution seems to fall into a pretty different category then your past CBM contributions.

Or am I missing the point?

Reply:
VICE has a repository of test programs which are tested against both emulation and real hardware.

Some of the code that I added to VICE was based on documentation about the actual hardware, and in most cases later verified by actual hardware owners to work as expected, sometimes not working as expected and therefor fixes were needed.

When writing emulation you need to start somewhere, even if it is just guess work based on the documentation, later tests done on the real hardware will then drive the emulation towards a matching behavior, however, if you don't even start on the emulation before having the actual hardware usually no incentive/motivation will be present for people to compare/test the emulation with the actual hardware.

In the case of drivers (which the code for cc65 is) imo you should be able to go by the documentation and/or other software that 'drives' the actual hardware, if an existing piece of code (like the setup/configuration disk of the hardware) drives the hardware you should be able to use a similar piece of code without needing to have the hardware. Maybe I'm too naive in this regard.

Maybe the code should be labeled experimental, and/or comments that the code is based on drivers for the actual hardware, and/or api documentation and no tests have been done on actual hardware, makes me wonder though, has for example the apple II machine detection code been tested on a macintosh iie option card, and ifnot, why is it in cc65 ?

Reply:
I understand the suggestion, I was trying to minimize the code size, and based on the documentation I didn't see any timing constraints.

Reply:
- Given the overall quality of the programmers documentation it doesn‘t mean anything that the documentation doesn‘t mention constraints.

- As mentioned before there are other places in the Apple II world where such constraints apply without being explictly documented.